zcash_primitives/src/transaction/sighash_v5.rs

```diff
@@ -0,0 +1,158 @@
+        } else {
+            self.txid_parts.transparent_digests.prevout_digest
+        };
+
+        let sequence_digest = if flag_anyonecanpay || flag_single || flag_none {
+            sequence_hash(&[])
+        } else {
+            self.txid_parts.transparent_digests.sequence_digest
+        };
+
+        let outputs_digest = if flag_single {
+            match self.signable_input {
+                SignableInput::Transparent { index, .. } if index < vout.len() => {
+                    outputs_hash(&[&vout[index]])
+                }
+                _ => outputs_hash::<TxOut>(&[]),
+            }
+        } else if flag_none {
+            outputs_hash::<TxOut>(&[])
+        } else {
+            self.txid_parts.transparent_digests.outputs_digest
+        };
+
+        let mut ch = HashWriter::new(ZCASH_TRANSPARENT_INPUT_HASH_PERSONALIZATION);
+        let per_input_digest = match self.signable_input {
+            SignableInput::Transparent {
+                index,
+                script_code,
+                value,
+            } => {
+                // ZIP-0143.10
+                // If we are serializing an input (i.e. this is not a JoinSplit signature hash):
+                //   a. outpoint (32-byte hash + 4-byte little endian)
+                //   b. scriptCode of the input (serialized as scripts inside CTxOuts)
+                //   c. value of the output spent by this input (8-byte little endian)
+                //   d. nSequence of the input (4-byte little endian)
+                let txin = &vin[index];
+                txin.prevout.write(&mut ch).unwrap();
+                script_code.write(&mut ch).unwrap();
+                ch.write_all(&value.to_i64_le_bytes()).unwrap();
+                ch.write_u32::<LittleEndian>(txin.sequence).unwrap();
+                ch.finalize()
+            }
+
+            _ => ch.finalize(),
+        };
+
+        TransparentDigests {
+            prevout_digest,
+            sequence_digest,
+            outputs_digest,
+            per_input_digest: Some(per_input_digest),
+        }
+    }
+
+    #[cfg(feature = "zfuture")]
+    fn digest_tze(&self, tze_inputs: &[TzeIn], _tze_outputs: &[TzeOut]) -> TzeDigests<Blake2bHash> {
+        let mut ch = HashWriter::new(ZCASH_TZE_INPUT_HASH_PERSONALIZATION);
+        let per_input_digest = match self.signable_input {
+            SignableInput::Tze {
+                index,
+                precondition,
+                value,
+            } if self.txversion == TxVersion::ZFuture => {
+                let tzein = &tze_inputs[index];
+
+                tzein.prevout.write(&mut ch).unwrap();
+                CompactSize::write(&mut ch, precondition.extension_id.try_into().unwrap()).unwrap();
+                CompactSize::write(&mut ch, precondition.mode.try_into().unwrap()).unwrap();
+                Vector::write(&mut ch, &precondition.payload, |w, e| w.write_u8(*e)).unwrap();
+                ch.write_all(&value.to_i64_le_bytes()).unwrap();
```

Comment:
@str4d now that I'm revisiting this signature hash code that we originally worked on together, I'm puzzled here - why is there a value associated with the TZE input?

Reply:
The logic for adding it was the same as for transparent inputs: without it, there's nothing in the spend that directly commits to the value of the spent output, so a hardware wallet can't verify the fee without being provided those input coins. By having the value be part of the signature digest, the hardware wallet can present accurate information without needing the coin being spent, and the resulting signature can only be used if the client wasn't lying about the coin's value.

This is also the thing I was saying we should check against the hardware wallet fee vulnerability from last year, which IIRC said that this defense was ineffective if the adversary could ask the user to sign multiple times.

Reply:
Ah, okay. I understand this better now than the last time we had this dicussion.